CONTRIBUTING.md

```diff
@@ -218,3 +218,125 @@ const entrypoints = {
+
+If you are contributing to this code base, then you need to be familiar with some of the underlying concepts that power the `Graph` class - LangGraph's main entrypoint. These concepts are intentionally not documented in the LangGraph docs because users of LangGraph do not need to understand them. This knowledge is exclusively for contributors.
+
+#### Pregel
+
+Let's start with Pregel. [Pregel](https://research.google/pubs/pregel-a-system-for-large-scale-graph-processing/) is an API for building graphs. 
+
+Some key concepts: 
+- Pregel graphs take an `input` and `output` as parameters which represent where the graph starts and ends
+- Pregel graphs take a mapping of nodes represented as `{nodeName: node}`
+- Each node subscribes to (one or more) channels. This defines when a node executes. Specifically, for a given `node N` that subscribes to `channel M`, whenever the _value_ of `channel M` changes, `node N` must be executed. Intuitively, it represents what the current node is _dependent_ on. 
+- Each node writes to (one or more) channels. This defines where the final value after a node is executed is stored, because nodes don't store their own value.
+- More on channels below. 
+
+To form an intuition around Pregel graphs, lets look at the example tests. 
+
+In the example below, the pregel graph is defined to start at `inputChannelName` and end at `outputChannelName`. The graph has a single node called `nodeOne` that transforms the input value by adding one to it. When the graph is invoked with an input value of `2`:
+1. `inputChannelName` gets set to a value of `2`, because it is defined as the input channel.
+2. Since `nodeOne` subscribes to `inputChannelName`, `nodeOne` executed. 
+3. `nodeOne` transforms `2` to `3` and `3` gets written to `outputChannelName`. 
+4. Since `outputChannelName` is defined as the graph's output, the execution ends and returns `3`.
+
+```ts
+  const addOne = jest.fn((x: number): number => x + 1);
+  const chain = Channel.subscribeTo("inputChannelName")
+    .pipe(addOne)
+    .pipe(Channel.writeTo("outputChannelName"));
+
+  const app = new Pregel({
+    nodes: { nodeOne: chain },
+    input: ["inputChannelName"],
+    output: ["outputChannelName"],
+  });
+
+  expect(await app.invoke({ input: 2 })).toEqual({ output: 3 });
+```
+
+This was a simple example, let's look at a more complicated example. 
+
+In the example below, the graph has one node. The `checkpointer` parameter in Pregel means that it persists the state at every step. If a checkpointer is specified, then `threadId` must be specified every time the graph is invoked and it represents the unique id of that invocation. 
+
+Invocation 1: 
+1. When the graph is invoked with `2`, `input` channels value becomes `2` 
+2. Node `one` runs because it is subscribed to `input`. The node tranforms `2` to `2` by running `inputPlusTotal`. 
+3. The value of channels `output` and `total` get set to `2` because node `one` writes to both channels 
```

Comment:
There is no `total` referenced above?

Reply:
oh the description is above the example 
i can change that if that's confusing